### PR TITLE
arch: arm: aarch32:cortex_m: timing.c: cortex M7 may need DWT unlock

### LINKED ISSUE #30322: tests: benchmarks: latency_measure: timing measurement values are all 0
**Describe the bug**
Running tests/benchmarks/latency_measure on nucleo_f746zg,
test is passed, but all timing measurement values are `0 ns`.

**To Reproduce**
1. west build -p auto -b nucleo_f746zg tests/benchmarks/latency_measure
2. west flash
3. See log

**Expected behavior**
Non null timing measurement value

**Logs and console output**
```
*** Booting Zephyr OS build zephyr-v2.4.0-1849-gc0650217cc7c  ***
START - Time Measurement
Timing results: Clock frequency: 72 MHz
Average thread context switch using yield                   :       0 cycles ,        0 ns
Average context switch time between threads (coop)          :       0 cycles ,        0 ns
Switch from ISR back to interrupted thread                  :       0 cycles ,        0 ns
Time from ISR to executing a different thread               :       0 cycles ,        0 ns
Time to create a thread (without start)                     :       0 cycles ,        0 ns
Time to start a thread                                      :       0 cycles ,        0 ns
Time to suspend a thread                                    :       0 cycles ,        0 ns
Time to resume a thread                                     :       0 cycles ,        0 ns
Time to abort a thread (not running)                        :       0 cycles ,        0 ns
Average semaphore signal time                               :       0 cycles ,        0 ns
Average semaphore test time                                 :       0 cycles ,        0 ns
Semaphore take time (context switch)                        :       0 cycles ,        0 ns
Semaphore give time (context switch)                        :       0 cycles ,        0 ns
Average time to lock a mutex                                :       0 cycles ,        0 ns
Average time to unlock a mutex                              :       0 cycles ,        0 ns
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Environment (please complete the following information):**
 - OS: Windows
 - Toolchain gnuarmemb
 - Commit SHA 1b65fd066f66596177fbce2f60f0deb2697bb2cb

**Additional context**

### PR DESCRIPTION
**Pull request by ABOSTM**
arch: arm: aarch32:cortex_m: timing.c: cortex M7 may need DWT unlock

On Cortex M7, we need to check the optional presence of
Lock Access Register (LAR) which is indicated in
Lock Status Register (LSR).
When present, a special access token must be written to unlock DWT
registers.

Fixes #30322

### CODE REVIEW

File: arch/arm/core/aarch32/cortex_m/timing.c [hunk 1]
```diff
@@ -14,6 +14,45 @@
 #include <timing/timing.h>
 #include <arch/arm/aarch32/cortex_m/cmsis.h>
 
+/* Define DWT LSR masks which are currentuly not defined by the CMSIS V5.1.2.
```
Comment (erwango) [inline arch/arm/core/aarch32/cortex_m/timing.c:17]:
```suggestion
/* Define DWT LSR masks which are currently not defined by the CMSIS V5.1.2.
```

File: arch/arm/core/aarch32/cortex_m/timing.c [hunk 2]
```diff
@@ -14,6 +14,45 @@
...
+				DWT->LAR = 0xC5ACCE55;
+			}
+		} else {
+			if ((lsr & DWT_LSR_Access_Msk) == 0) {
+				/* Acess is unlocked. Lock it */
+				DWT->LAR = 0;
+			}
+		}
+	}
+#else /* __CORTEX_M */
```
Comment (erwango) [inline arch/arm/core/aarch32/cortex_m/timing.c:52]:
```suggestion
#else /* CONFIG_CPU_CORTEX_M7 */
```

File: arch/arm/core/aarch32/cortex_m/timing.c [hunk 3]
```diff
@@ -14,6 +14,45 @@
...
+	UNUSED(ena);
+#endif /* __CORTEX_M */
```
Comment (erwango) [inline arch/arm/core/aarch32/cortex_m/timing.c:54]:
```suggestion
#endif /* CONFIG_CPU_CORTEX_M7 */
```

File: arch/arm/core/aarch32/cortex_m/timing.c [hunk 4]
```diff
@@ -14,6 +14,45 @@
...
+			}
+		} else {
+			if ((lsr & DWT_LSR_Access_Msk) == 0) {
+				/* Acess is unlocked. Lock it */
+				DWT->LAR = 0;
+			}
+		}
+	}
+#else /* __CORTEX_M */
+	UNUSED(ena);
```
Comment (erwango) [inline arch/arm/core/aarch32/cortex_m/timing.c:53]:
```suggestion
	ARG_UNUSED(ena);
```

**Review by erwango — CHANGES_REQUESTED** [review]

File: arch/arm/core/aarch32/cortex_m/timing.c (hunk 1 above)
Comment (ABOSTM) [reply arch/arm/core/aarch32/cortex_m/timing.c]: Done

File: arch/arm/core/aarch32/cortex_m/timing.c (hunk 2 above)
Comment (ABOSTM) [reply arch/arm/core/aarch32/cortex_m/timing.c]: Done

File: arch/arm/core/aarch32/cortex_m/timing.c (hunk 3 above)
Comment (ABOSTM) [reply arch/arm/core/aarch32/cortex_m/timing.c]: Done

File: arch/arm/core/aarch32/cortex_m/timing.c (hunk 4 above)
Comment (ABOSTM) [reply arch/arm/core/aarch32/cortex_m/timing.c]: Done

**Comment by ABOSTM** [comment]
This patch is inspired from Arduino code (Copyright STMicroelectronics) with BSD-3-Clause license, compatible with Apache License.
https://github.com/stm32duino/Arduino_Core_STM32/blob/5f95b95cd5c8f7673886ca0d243684421f4595b8/libraries/SrcWrapper/src/stm32/dwt.c#L54

**Review by erwango — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]
Looks ok to me, thanks @ABOSTM 

Interesting that LSR seems to only be implemented, optionally, in CM7, while LSR register is available in all other core variant implementations.

**Pull request merged**